src/python/pants/backend/scala/lint/scalafmt/rules.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import textwrap
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.scala.lint.scala_fmt import ScalaFmtRequest
+from pants.backend.scala.lint.scalafmt.skip_field import SkipScalafmtField
+from pants.backend.scala.lint.scalafmt.subsystem import ScalafmtSubsystem
+from pants.backend.scala.target_types import ScalaSourceField
+from pants.core.goals.fmt import FmtResult
+from pants.core.goals.lint import LintRequest, LintResult, LintResults
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import BashBinary, FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import MaterializedClasspath, MaterializedClasspathRequest
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.testutil.rule_runner import logging
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class ScalafmtFieldSet(FieldSet):
+    required_fields = (ScalaSourceField,)
+
+    source: ScalaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipScalafmtField).value
+
+
+class ScalafmtRequest(ScalaFmtRequest, LintRequest):
+    field_set_type = ScalafmtFieldSet
+
+
+class ScalafmtToolLockfileSentinel(JvmToolLockfileSentinel):
+    options_scope = ScalafmtSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: ScalafmtRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: Process
+    original_digest: Digest
+
+
+@logging
+@rule(level=LogLevel.DEBUG)
+async def setup_google_java_format(
+    setup_request: SetupRequest,
+    tool: ScalafmtSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
```

Comment:
(unrelated: we could probably embed the `BashBinary` in the `jdk_setup` to avoid the extra import/arg)